### PR TITLE
Add `preferExplicitFalse` rule

### LINKED ISSUE #2356: Rule idea: prefer `== false` over `!` prefix negation
### Description

A new opt-in rule that converts prefix `!` negation to explicit `== false` comparison.

```diff
- if !isHidden {
+ if isHidden == false {

- guard !array.isEmpty else { return }
+ guard array.isEmpty == false else { return }

- while !finished {
+ while finished == false {
```

Some teams prefer `== false` for readability, as the `!` prefix can be easy to miss when scanning code.

### Why this is safe

If `!expr` compiles, `expr` is guaranteed to be a non-optional `Bool` (applying `!` to `Bool?` is a compile error), so `expr == false` is always semantically equivalent. No type information needed.

### Suggested behavior

| Before | After |
|---|---|
| `!flag` | `flag == false` |
| `!array.isEmpty` | `array.isEmpty == false` |
| `!foo.bar()` | `foo.bar() == false` |
| `!(a && b)` | `(a && b) == false` |

For the `!(expr)` case, the outer parens around the group are preserved and `!` is removed.

### Related

This is the reverse direction of #2148, which was declined because `== false` → `!` requires type information to handle `Bool?` safely. That concern does not apply in this direction.

### PR DESCRIPTION
**Pull request by KYHyeon**
## Summary

- Add new opt-in rule `preferExplicitFalse` that converts prefix `!` negation to explicit `== false` comparison
- The transformation is always safe: if `!expr` compiles, `expr` is guaranteed `Bool`, so `expr == false` is semantically equivalent
- Disabled by default (`disabledByDefault: true`), enabled via `--enable preferExplicitFalse`

### Examples

| Before | After |
|---|---|
| `!flag` | `flag == false` |
| `!array.isEmpty` | `array.isEmpty == false` |
| `!foo.bar()` | `foo.bar() == false` |
| `!(a && b)` | `(a && b) == false` |

Closes #2356

## Test plan

- [x] 30 test cases covering basic negation, property/method chains, subscripts, force unwrap, closures, where clauses, array/dictionary literals, and no-change cases
- [x] Rule tests pass: `./Scripts/test_rule.sh preferExplicitFalse`
- [x] Full test suite passes: `swift test` (5154 tests, 0 failures)

### CODE REVIEW

**Comment by calda** [comment]
I ran the rule on Airbnb's codebase and found cases where it breaks the build. Most of these cases are trailing closures, like:

```swift
!myArray.contains {
  $0 == value
}
```

which is being auto-corrected to:

```swift
myArray.contains == false {
  $0 == value
}
```

I can also think of examples where this would create operator precedence ambiguity and cause build errors:

```swift
let a = true
let b = false

// Compiles
print(!a == b)

// Error: adjacent operators are in non-associative precedence group 'ComparisonPrecedence'
print(a == false == b)

// Error: adjacent operators are in non-associative precedence group 'ComparisonPrecedence'
print(a == false != b)
```

File: Sources/Rules/PreferExplicitFalse.swift [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+}
+
+// MARK: - Helpers
+
+extension Formatter {
+    /// Finds the end of the postfix expression starting at `index`, which is the
+    /// first non-space token after a prefix `!`. This includes member access chains,
+    /// method calls, subscripts, and postfix operators, but stops before infix
+    /// operators, `is`/`as`, trailing closures, or linebreaks.
+    func endOfPrefixNegationOperand(startingAt index: Int) -> Int? {
```
Comment (calda) [inline Sources/Rules/PreferExplicitFalse.swift]: Can we use the existing `parseExpressionRange` helper? It should already handle all of these cases and more (e.g. trailing closures, multiple trailing closures, more complex method call / access chains)

**Comment by KYHyeon** [comment]
> let a = true
> let b = false
> 
> // Compiles
> print(!a == b)
> 
> // Error: adjacent operators are in non-associative precedence group 'ComparisonPrecedence'
> print(a == false == b)
> 
> // Error: adjacent operators are in non-associative precedence group 'ComparisonPrecedence'
> print(a == false != b)

We could transform `!a == b` to `(a == false) == b` to avoid the non-associative operator chain, but since the goal of this  
  rule is to improve readability, that doesn't seem like a great trade-off.                                                
                                                                                                                           
  Would it make sense to skip the transformation in this case instead? This would be similar to how yodaConditions handles 
  unsafe cases in https://github.com/nicklockwood/SwiftFormat/issues/699 — where the rule opts out when the transformation could break the build.

File: Sources/Rules/PreferExplicitFalse.swift (hunk 1 above)
Comment (KYHyeon) [reply Sources/Rules/PreferExplicitFalse.swift]: Great suggestion! I've updated the implementation to use parseExpressionRange. One thing to note is that parseExpressionRange includes infix operators in its returned range, but prefix ! binds tighter than any infix operator.
  So I still needed a scan to find the boundary before the first infix operator.                                           
   
  The trailing closure issue you found is also resolved naturally with this approach — parseExpressionRange includes the trailing closure in the expression range, and since {...} is a scope (not an infix operator), the scan skips over it   correctly.

  I also added a guard to skip the transformation when followed by a comparison operator (==, !=, etc.) to avoid the  non-associative ComparisonPrecedence chain error you mentioned.

  Thanks for the thorough review and for testing on the Airbnb codebase!

https://github.com/nicklockwood/SwiftFormat/pull/2357/commits/67a25dc97738c9c3d56d6bddaa80edc56c3b789b

**Comment by calda** [comment]
Either works for me. Skipping in that case is easiest since this is a more niche edge case.

**Comment by calda** [comment]
Ran on our codebase again and found another issue we need to handle: `a == !b` can't become `a == b == false` (so, the `==` operator could be before or after the subexpression)

File: Sources/Rules/PreferExplicitFalse.swift [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+            startingAt: index
+        ) else {
+            return nil
+        }
+
+        // parseExpressionRange includes infix operators in the expression range,
+        // but `!` binds tighter than any infix operator. Scan through the range
+        // and stop at the first infix operator (excluding member access `.`).
+        var i = index
+        while i <= expressionRange.upperBound {
```
Comment (calda) [inline Sources/Rules/PreferExplicitFalse.swift]: If you just do `formatter.index(of: .operator, in: index ... expressionRange.upperBound)` (etc), that already skips over any scopes that it finds (you wouldn't need to manually skip over intermediate scopes)

Comment (KYHyeon) [reply Sources/Rules/PreferExplicitFalse.swift]: Applied the suggested approach of using Formatter's traversal APIs instead of manual scope handling.

  One note: I used `index(in:where:)` instead of `index(of: .operator, in:, if:)` because the latter only checks the if condition on the first .operator token it finds — if that fails, it returns `nil` without continuing to search.
                                                                                                                                                           
  For example, in `!foo! == bar, index(of: .operator, in:, if: { $0.isOperator(ofType: .infix) })` would match the postfix `!` first, fail the infix check, and return `nil` — missing `==` entirely. `index(in:where:)` evaluates the full condition on each candidate, so it correctly skips the postfix ! and finds `==`.

  Also addressed the `a == !b` case you found, and while adding tests for it I discovered a few more edge cases `(!foo is Bar, !foo as? Bar, !foo! == bar)` that needed handling as well. Thanks for the thorough testing on your codebase!

https://github.com/nicklockwood/SwiftFormat/pull/2357/commits/b20193f229aae670a2705d6f84cc24c77fcab7b9

**Comment by nicklockwood** [comment]
Can't `!a == b` be more naturally expressed as `a != b` ?

**Comment by calda** [comment]
True! I did find examples of thinks like `!a == b` and `a != !b` in our codebase though.

File: Sources/Rules/PreferExplicitFalse.swift [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+            }
+
+            if formatter.currentScope(at: notIndex) == .startOfScope("#if") {
+                return
+            }
+
+            // Skip if adjacent to a comparison or casting operator —
+            // inserting `== false` would create a non-associative chain
+            // or change precedence. e.g., `a == !b` → `a == b == false`
+            if let prevIndex = formatter.index(of: .nonSpaceOrCommentOrLinebreak, before: notIndex),
```
Comment (calda) [inline Sources/Rules/PreferExplicitFalse.swift:37]: I think preserving these as-is is the right call
